asllib/doc/ASLTypingReference.tex

```diff
@@ -8694,3822 +8837,7477 @@ \subsection{Comments}
+  \item TypingRule.BuildDependencies (see \secref{TypingRule.BuildDependencies})
+  \item TypingRule.DeclDependencies (see \secref{TypingRule.DeclDependencies})
+  \item TypingRule.TypeCheckMutuallyRec (see \secref{TypingRule.TypeCheckMutuallyRec})
+  \item TypingRule.FoldEnvAndFs (see \secref{TypingRule.FoldEnvAndFs})
+  \item TypingRule.DefDecl (see \secref{TypingRule.DefDecl})
+  \item TypingRule.DefEnumLabels (see \secref{TypingRule.DefEnumLabels})
+  \item TypingRule.UseDecl (see \secref{TypingRule.UseDecl})
+  \item TypingRule.UseTy (see \secref{TypingRule.UseTy})
+  \item TypingRule.UseSubtypes (see \secref{TypingRule.UseSubtypes})
+  \item TypingRule.UseExpr (see \secref{TypingRule.UseExpr})
+  \item TypingRule.UseLexpr (see \secref{TypingRule.UseLexpr})
+  \item TypingRule.UsePattern (see \secref{TypingRule.UsePattern})
+  \item TypingRule.UseSlice (see \secref{TypingRule.UseSlice})
+  \item TypingRule.UseBitfield (see \secref{TypingRule.UseBitfield})
+  \item TypingRule.UseConstraint (see \secref{TypingRule.UseConstraint})
+  \item TypingRule.UseStmt (see \secref{TypingRule.UseStmt})
+  \item TypingRule.UseLDI (see \secref{TypingRule.UseLDI})
+  \item TypingRule.UseCase (see \secref{TypingRule.UseCase})
+  \item TypingRule.UseCatcher (see \secref{TypingRule.UseCatcher})
+\end{itemize}
+
+\section{TypingRule.TypeCheckAST \label{sec:TypingRule.TypeCheckAST}}
+\hypertarget{def-typecheckast}{}
+The relation
+\[
+\typecheckast(
+  \overname{\staticenvs}{\tenv} \aslsep
+  \overname{\decl^*}{\decls}
+) \aslrel
+(\overname{\decl^*}{\newdecls} \times \overname{\staticenvs}{\newtenv})
+\cup \overname{\TTypeError}{\TypeErrorConfig}
+\]
+annotates a list of declarations $\decls$ in an input static environment $\tenv$,
+yielding an output static environment $\newtenv$ and annotated list of declarations $\newdecls$.
+Otherwise, the result is a type error.
+
+\begin{definition}[Strongly Connected Components]
+\hypertarget{def-scc}{}
+Given a graph $G=(V, E)$, a \\ subset of its nodes $C \subseteq V$ is called
+a \emph{strongly connected component} of $G$ if
+every pair of nodes $u,v \in C$ reachable from one another.
+
+The \emph{strongly connected components} of a graph $(V, E)$ uniquely partitions its set of
+nodes $V$ into a set of strongly connected components:
+\[
+\SCC(V, E) \triangleq \{ C \subseteq V \;|\; \forall u,v\in C.\ (u,v), (v,u) \in E^* \} \enspace.
+\]
+
+We write $E^*$ to denote the reflexive-transitive closure of $E$.
+\end{definition}
+
+\begin{definition}[Topological Ordering]
+For a graph $G=(V, E)$ and its \\
+strongly connected components $\comps \triangleq \SCC(V, E)$,
+we say that $C_1\in\comps$ is ordered before $C_2\in\comps$, denoted $C_1 < C_2$,
+if the following condition holds:
+\[
+C_1 < C_2 \Leftrightarrow \exists c_1\in C_1.\ c_2\in C_2.\ (c_1,c_2) \in E^* \enspace.
+\]
+This ordering is not total. That is, there may very well exist to strongly connected components
```

Comment:
there may very well exist to strongly connected components -> there may exist two strongly connected components